src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeoAuthProvider.scala

```diff
@@ -0,0 +1,37 @@
+package org.broadinstitute.dsde.workbench.leonardo.model
+
+import java.util.UUID
+
+import org.broadinstitute.dsde.workbench.leonardo.api.AuthorizationError
+import org.broadinstitute.dsde.workbench.model.WorkbenchUserEmail
+
+import scala.concurrent.Future
+
+sealed trait ProjectAction
+case object ListClusters extends ProjectAction
+case object CreateClusters extends ProjectAction
+
+sealed trait NotebookClusterAction
+case object GetClusterDetails extends NotebookClusterAction
+case object ConnectToCluster extends NotebookClusterAction
+case object LocalizeDataToCluster extends NotebookClusterAction
+case object DestroyCluster extends NotebookClusterAction
+
+abstract class LeoAuthProvider {
+  //Does this user have permission in all projects to perform this action?
+  def hasPermissionInAllProjects(user: WorkbenchUserEmail, action: ProjectAction): Future[Boolean]
+
+  //Does the user have permission in this project to perform this action?
+  def hasProjectPermission(user: WorkbenchUserEmail, action: ProjectAction, project: GoogleProject): Future[Boolean]
+
+  //List the projects in which the user has ListClusters action.
+  //If hasPermissionInAllProjects(_, ListClusters) returns true in your implementation, you should return Seq.empty() for this function.
+  def getProjectsWithListClustersPermission(user: WorkbenchUserEmail): Future[Set[GoogleProject]]
```

Comment:
This API seems a little bit odd. What if instead we made the "list clusters" endpoint take a Google project? (We could have a special admin-only endpoint for listing all clusters.) Then it would be covered by the `hasProjectPermission` method and we wouldn't need `getProjectsWithListClustersPermission`, I think. Isn't the normal FC use case to list clusters in a single project anyway? Is there ever a use case for a multi-project list (outside of admin use)?

Reply:
+1 -- it's also ambiguous what "All" means for projects?  All in the universe?  All in the leo world, etc.  If we can break it into "admin" and "project" permissions, that might work well.

also, should Leo know about "Workbench" (e.g. WorkbenchUserEmail).  Is that just an email or are there special constraints?  Thinking about our verily friends

Reply:
Agreed it's strange. But with my user hat on, why _shouldn't_ I be able to search on everything I can see? Maybe google project is less interesting to me than some other label that I've added. Maybe I want to look up all the clusters that have `project: diabetes` so I can find out where all my data is.

Reply:
@kcibul All in the Leo world. We can rename it if that'd be better.
Also -- `WorkbenchUserEmail` is just an email. I believe this is being renamed to `WorkbenchEmail` soon. Easy enough to make it a String, though I'd say "Leo is a Workbench service" is fine (as opposed to "Leo is part of FireCloud", which is not)

Reply:
The `hasPermissionInAllProjects` thing is for situations like whitelist-based auth, which says "yes" if you are on an email list. But it doesn't know anything about project-level permissions, so if you ask it "what projects can I spin up compute in?" (which seems like a reasonable question to ask an auth provider), it wants to say "everything" but there's no way of representing that in a list-of-projects.